### PR TITLE
Feature/initial

### PR DESCRIPTION
**Pull request by djkwagala**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,33 @@
...
+[![codecov](https://codecov.io/gh/mendixlabs/list-view-controls/branch/master/graph/badge.svg)](https://codecov.io/gh/mendixlabs/list-view-controls)
+
+# List view controls
+Enable users to filter or search a list view.
+
+## Available widgets
+* `Check box filter`
+* `Drop down filter`
+* `Drop down sort`
+* `Text box search`
```
Comment (Andries-Smit) [inline README.md]: Remind that this README is going to be the app store readme page, so make a short 1 line description what the widget does. Move the other readme files into folder `docs` and dont start there name with REAMDE

Add a single screen shot with all different widgets in there, as Danny if he has one.

File: README.md [hunk 2]
```diff
@@ -1 +1,33 @@
...
+
+## Demo projects
+[https://listviewcontrols.mxapps.io](https://listviewcontrols.mxapps.io)
```
Comment (Andries-Smit) [inline README.md]: Create a task for Jeroen, asking for a singe appstore icon for the all widget combined

File: development.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
+# Development and contribution
+Prerequisite: Install git, node package manager, webpack CLI, grunt CLI, Karma CLI
```
Comment (Andries-Smit) [inline development.md]: move into directory `docs` with the other widget readme files
Always keep the main readme in the root, for GitHub to pick it up as the main documentation

File: package.json [hunk 4]
```diff
@@ -0,0 +1,105 @@
+{
+  "name": "list-view-controls",
+  "widgetName": "ListViewControls",
+  "version": "1.0.0",
+  "description": "Filter Mendix list views",
```
Comment (Andries-Smit) [inline package.json]: update the generic description

File: src/CheckBoxFilter/Validate.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
+import { ContainerProps } from "./components/CheckBoxFilterContainer";
```
Comment (Andries-Smit) [inline src/CheckBoxFilter/Validate.ts:1]: 👍 good to put in a file on its own

File: src/CheckBoxFilter/Validate.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        }
+        if (!props.isWebModeler && window.mx.isOffline() && props.filterBy === "XPath") {
+            errorMessage.push("Filter by 'XPath' is not supported in offline mode");
+        }
+        if (!props.isWebModeler && window.mx.isOffline() && props.unCheckedFilterBy === "XPath") {
+            errorMessage.push("Unchecked filter by 'XPath' is not supported in offline mode");
+        }
+        if (!props.isWebModeler && !props.mxObject && props.filterBy === "XPath" && props.constraint.indexOf("[%CurrentObject%]'") > -1) {
+            errorMessage.push("Requires a context object");
+        }
```
Comment (Andries-Smit) [inline src/CheckBoxFilter/Validate.ts]: maybe group the `! porps.isWebmodeler` into a singe statement and the in the sub do the other checks

``` ts
if (!props.isWebModeler) {
        if (window.mx.isOffline() && props.filterBy === "XPath") {
             errorMessage.push("Filter by 'XPath' is not supported in offline mode");
         }
         if (window.mx.isOffline() && props.unCheckedFilterBy === "XPath") {
             errorMessage.push("Unchecked filter by 'XPath' is not supported in offline mode");
         }
         if (!props.mxObject && props.filterBy === "XPath" && props.constraint.indexOf("[%CurrentObject%]'") > -1) {
             errorMessage.push("Requires a context object");
    }
}
```

**Review by Andries-Smit — COMMENTED** [review]
more to come

File: karma.conf.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        });
+    }
+
+    const configuration = {
+        basePath: "",
+        frameworks: [ "jasmine" ],
+        files: [
+            { pattern: "src/**/*.ts", watched: true },
+            { pattern: "tests/**/*.ts", watched: true },
+            "tests/test-index.js"
```
Comment (Andries-Smit) [inline karma.conf.js:36]: Missing someting.. only 13 test are  running. are there more?
``` txt
DropDownFilter
renders the structure correctly
select
changes value
updates when the select option changes
Dropdown
renders the structure correctly
renders with the specified default sort
select
changes value
updates when the select option changes
TextBoxSearch
renders the structure correctly
renders with the specified placeholder
input
accepts value
renders with specified default query
updates when the search value changes
is cleared when the remove button is clicked
```

File: README.DropDownFilter.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
+# Drop down filter
+Enable users to filter a list view at run time, with various options that can be selected from a drop-down
+
+## Features
+* Filter items by an attribute
+* Filter items by XPath
+* Select a filter from a list of options
+* Set a default filter option
+
+## Dependencies
```
Comment (Andries-Smit) [inline README.DropDownFilter.md]: no need to specify dependency per widget, should only be in the general readme

File: README.DropDownFilter.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+Mendix 7.6
+
+## Demo project
```
Comment (Andries-Smit) [inline README.DropDownFilter.md]: Not `Demo project` => `Sample`.
Section demo project is for the URL to the sandbox
Please update for versions

File: README.DropDownFilter.md [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+When filtering by `XPath`, input a constraint to filter by.
+`None` is for an empty option which resets the filter then selected.
+
+**NB: The `None` filter option should always appear at the top of the list and it does not require a caption.** 
+**Also to note, only one empty filter option should be selected.**
+
+![XPathConstraint](assets/DropDownFilter/XPathConstraint.png)
+
+## Issues, suggestions and feature requests
+Please report issues at [https://github.com/mendixlabs/listview-filter/issues](https://github.com/mendixlabs/listview-filter/issues).
```
Comment (Andries-Smit) [inline README.DropDownFilter.md]: issue only the over readme, (count ver all widgets)

File: README.DropDownFilter.md [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+## Development and contribution
```
Comment (Andries-Smit) [inline README.DropDownFilter.md]: Dev section only for the root README file (remove from all widgets)

File: README.DropDownSort.md [hunk 12]
```diff
@@ -0,0 +1,38 @@
+
```
Comment (Andries-Smit) [inline README.DropDownSort.md]: file should not be starting with new line

File: src/Shared/DataSourceHelper/DataSourceHelper.ts [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+            if (this.requiresUpdate) {
+                this.requiresUpdate = false;
+                this.iterativeUpdateDataSource();
+            } else {
+                this.updateInProgress = false;
+            }
+        });
+    }
+
+    private versionCompatibility(version: Version, widgetId: string) {
```
Comment (Andries-Smit) [inline src/Shared/DataSourceHelper/DataSourceHelper.ts:88]: Can we simplify this, by just checking the  version, there is low/no change that some other widget will do the  same thing....

File: src/TextBoxSearch/components/TextBoxSearchContainer.ts [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+            DataSourceHelper.hideContent(targetNode);
+        }
+    }
+
+    componentDidUpdate(_prevProps: ContainerProps, prevState: ContainerState) {
+        if (this.state.listViewAvailable && !prevState.listViewAvailable) {
+            this.applySearch(this.props.defaultQuery);
+        }
+    }
+    render() {
```
Comment (Andries-Smit) [inline src/TextBoxSearch/components/TextBoxSearchContainer.ts]: add white line before render

Comment (Andries-Smit) [inline src/TextBoxSearch/components/TextBoxSearchContainer.ts]: missing return type, all function that not return void should declare their run time

File: src/TextBoxSearch/components/TextBoxSearchContainer.ts [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+
+    componentDidMount() {
+        const filterNode = findDOMNode(this).parentNode as HTMLElement;
+        const targetNode = SharedUtils.findTargetNode(filterNode);
+        if (targetNode) {
+            DataSourceHelper.hideContent(targetNode);
+        }
+    }
+
+    componentDidUpdate(_prevProps: ContainerProps, prevState: ContainerState) {
```
Comment (Andries-Smit) [inline src/TextBoxSearch/components/TextBoxSearchContainer.ts]: no acronyms... `_previousProps` and `previousState`

File: src/TextBoxSearch/components/TextBoxSearchContainer.ts [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+                onTextChange: this.applySearch,
+                placeholder: this.props.placeHolder
+            });
+        }
+
+        return null;
+    }
+
+    private applySearch(searchQuery: string) {
+        // construct constraint based on search query
```
Comment (Andries-Smit) [inline src/TextBoxSearch/components/TextBoxSearchContainer.ts:90]: comments should start with capital

File: src/TextBoxSearch/components/TextBoxSearchContainer.ts [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+    private getConstraint(searchQuery: string) {
+        const { targetListView } = this.state;
+        const constraints: string[] = [];
+
+        if (targetListView && targetListView._datasource) {
+            this.props.attributeList.forEach(searchAttribute => {
+                constraints.push(`contains(${searchAttribute.attribute},'${searchQuery}')`);
+            });
+
+            return "[" + constraints.join(" or ") + "]";
```
Comment (Andries-Smit) [inline src/TextBoxSearch/components/TextBoxSearchContainer.ts:107]: or interpolate 
``` ts
 return "[" + constraints.join(" or ") + "]";
 return `[${constraints.join(" or ")}]`; 
```
not sure what looks nicers

File: tests/e2e/CheckboxFilter.spec.ts [hunk 18]
```diff
@@ -0,0 +1,52 @@
+import defaultFilter from "./CheckBoxFilter/pages/default.page";
+
+describe("CheckBoxFilter", () => {
```
Comment (Andries-Smit) [inline tests/e2e/CheckboxFilter.spec.ts]: Tests are written like sentences. And lowest level should start with a Capitall

File: tests/e2e/DropDownFilter.spec.ts [hunk 19]
```diff
@@ -0,0 +1,13 @@
+import defaultFilter from "./DropDownFilter/pages/default.page";
+
+const dropDownTestValue = "Uganda";
+describe("DropDownFilter", () => {
```
Comment (Andries-Smit) [inline tests/e2e/DropDownFilter.spec.ts]: `Drop down filter`

File: tests/e2e/CheckboxFilter.spec.ts (hunk 18 above)
Comment (Andries-Smit) [reply tests/e2e/CheckboxFilter.spec.ts]: `Check box filter`

File: tests/e2e/DropDownSort.spec.ts [hunk 20]
```diff
@@ -0,0 +1,14 @@
+import page from "./DropDownSort/pages/home.page";
+
+const testValue = "Red";
+
+describe("Dropdown sort", () => {
```
Comment (Andries-Smit) [inline tests/e2e/DropDownSort.spec.ts]: `Drop down sort`

File: tests/e2e/TextBoxSearch.spec.ts [hunk 21]
```diff
@@ -0,0 +1,40 @@
+import page from "./TextSearch/pages/home.page";
+import { Element } from "webdriverio";
+
+const testValue = "Uganda";
+
+describe("Search", () => {
```
Comment (Andries-Smit) [inline tests/e2e/TextBoxSearch.spec.ts]: `Text ox search`

File: tsconfig.json [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+      "skipLibCheck": false,
+      "noUnusedLocals": true,
+      "noUnusedParameters": true
+    },
+    "exclude": [
+      "dist/",
+      "node_modules/"
+    ]
+  }
+
```
Comment (Andries-Smit) [inline tsconfig.json]: Always have new line at end of file

**Review by Andries-Smit — COMMENTED** [review]
First level of scanning code. Promising

File: README.DropDownFilter.md (hunk 8 above)
Comment (djkwagala) [reply README.DropDownFilter.md]: Done

File: README.DropDownFilter.md (hunk 9 above)
Comment (djkwagala) [reply README.DropDownFilter.md]: Done

File: README.DropDownFilter.md (hunk 10 above)
Comment (djkwagala) [reply README.DropDownFilter.md]: Done

File: README.DropDownFilter.md (hunk 11 above)
Comment (djkwagala) [reply README.DropDownFilter.md]: Done

File: src/TextBoxSearch/components/TextBoxSearchContainer.ts (hunk 16 above)
Comment (jseremba) [reply src/TextBoxSearch/components/TextBoxSearchContainer.ts:90]: Changed

File: tests/e2e/CheckboxFilter.spec.ts (hunk 18 above)
Comment (jseremba) [reply tests/e2e/CheckboxFilter.spec.ts]: Changed

File: tests/e2e/DropDownFilter.spec.ts (hunk 19 above)
Comment (jseremba) [reply tests/e2e/DropDownFilter.spec.ts]: Done

File: tests/e2e/DropDownSort.spec.ts (hunk 20 above)
Comment (jseremba) [reply tests/e2e/DropDownSort.spec.ts]: Done

File: tests/e2e/TextBoxSearch.spec.ts (hunk 21 above)
Comment (jseremba) [reply tests/e2e/TextBoxSearch.spec.ts]: Done

File: README.DropDownSort.md (hunk 12 above)
Comment (djkwagala) [reply README.DropDownSort.md]: Done

File: src/TextBoxSearch/components/TextBoxSearchContainer.ts (hunk 15 above)
Comment (djkwagala) [reply src/TextBoxSearch/components/TextBoxSearchContainer.ts]: Done

File: src/TextBoxSearch/components/TextBoxSearchContainer.ts (hunk 14 above)
Comment (djkwagala) [reply src/TextBoxSearch/components/TextBoxSearchContainer.ts]: I think the react's `render` function would be exceptional.
All the other functions with indeed have a return type.

File: tsconfig.json (hunk 22 above)
Comment (djkwagala) [reply tsconfig.json]: Done

File: src/TextBoxSearch/components/TextBoxSearchContainer.ts (hunk 14 above)
Comment (djkwagala) [reply src/TextBoxSearch/components/TextBoxSearchContainer.ts]: Done

File: src/CheckBoxFilter/Validate.ts (hunk 6 above)
Comment (djkwagala) [reply src/CheckBoxFilter/Validate.ts]: Done

File: package.json (hunk 4 above)
Comment (djkwagala) [reply package.json]: `Search and filter Mendix list views`

File: README.md (hunk 1 above)
Comment (djkwagala) [reply README.md]: Done

**Pull request merged**